bundles/org.openhab.binding.hue/src/main/java/org/openhab/binding/hue/internal/connection/Clip2Bridge.java

```diff
@@ -0,0 +1,980 @@
+     * which, in the case of fatal errors, gets completed exceptionally.
+     *
+     * It handles the following fatal error events by notifying the containing class:
+     *
+     * <li>onHeaders() HTTP unauthorized codes</li>
+     */
+    private abstract class BaseStreamListenerAdapter<T> extends Stream.Listener.Adapter {
+        protected final CompletableFuture<T> completable = new CompletableFuture<T>();
+        private String contentType = "UNDEFINED";
+
+        protected T awaitResult() throws ExecutionException, InterruptedException, TimeoutException {
+            return completable.get(TIMEOUT_SECONDS, TimeUnit.SECONDS);
+        }
+
+        protected void fatalError(Http2Error error) {
+            Exception e;
+            if (Http2Error.UNAUTHORIZED.equals(error)) {
+                e = new HttpUnauthorizedException("HTTP 2 request not authorized");
+            } else {
+                e = new ApiException("HTTP 2 stream " + error.toString().toLowerCase());
+            }
+            completable.completeExceptionally(e);
+            fatalErrorDelayed(this, error);
+        }
+
+        /**
+         * Return the HTTP content type.
+         *
+         * @return content type e.g. 'application/json'
+         */
+        protected String getContentType() {
+            return contentType;
+        }
+
+        /**
+         * Check the reply headers to see whether the request was authorised.
+         */
+        @Override
+        public void onHeaders(@Nullable Stream stream, @Nullable HeadersFrame frame) {
+            Objects.requireNonNull(frame);
+            MetaData metaData = frame.getMetaData();
+            if (metaData.isResponse()) {
+                Response responseMetaData = (Response) metaData;
+                int httpStatus = responseMetaData.getStatus();
+                switch (httpStatus) {
+                    case HttpStatus.UNAUTHORIZED_401:
+                    case HttpStatus.FORBIDDEN_403:
+                        fatalError(Http2Error.UNAUTHORIZED);
+                    default:
+                }
+                contentType = responseMetaData.getFields().get(HttpHeader.CONTENT_TYPE).toLowerCase();
+            }
+        }
+    }
+
+    /**
+     * Adapter for listening to regular HTTP 2 GET/PUT request stream events.
+     *
+     * It assembles the incoming text data into an HTTP 'content' entity. And when the last data frame arrives, it
+     * returns the full content by completing the CompletableFuture with that data.
+     *
+     * In addition to those handled by the parent, it handles the following fatal error events by notifying the
+     * containing class:
+     *
+     * <li>onIdleTimeout()</li>
+     * <li>onTimeout()</li>
+     */
+    private class ContentStreamListenerAdapter extends BaseStreamListenerAdapter<String> {
+        protected final List<String> contentResult = new ArrayList<>();
```

Comment:
On a second thought, since it's only about string concatanation, maybe using a `StringBuilder` is more appropriate?
-> https://github.com/maniac103/openhab2-addons/commit/dc235812d1d296893e108ea997847d7d2a5eaf7d

Reply:
Ok.